### PR TITLE
Added ability to define protocol version when serializing

### PR DESCRIPTION
**Pull request by DaOneLuna**

### CODE REVIEW

File: core/src/protocols/farmer.rs [hunk 1]
```diff
@@ -205,62 +241,72 @@ impl dg_xch_serialize::ChiaSerialize for DeclareProofOfSpace {
...
+                &self.foliage_transaction_block_data,
+                version,
+            ));
+        }
+        if version >= ChiaProtocolVersion::Chia0_0_36 {
+            bytes.extend(dg_xch_serialize::ChiaSerialize::to_bytes(
+                &self.rc_block_unfinished,
+                version,
+            ));
+        }
```
Comment (felixbrucker) [inline core/src/protocols/farmer.rs:283]: These can be grouped together in the same if block

File: core/src/protocols/harvester.rs [hunk 2]
```diff
@@ -68,72 +66,89 @@ impl From<u8> for SigningDataKind {
...
+                &self.farmer_reward_address_override,
+                version,
+            ));
+        }
+        if version >= ChiaProtocolVersion::Chia0_0_36 {
+            bytes.extend(dg_xch_serialize::ChiaSerialize::to_bytes(
+                &self.fee_info,
+                version,
+            ));
+        }
```
Comment (felixbrucker) [inline core/src/protocols/harvester.rs:124]: Same here

File: core/src/protocols/harvester.rs [hunk 3]
```diff
@@ -150,52 +165,65 @@ impl dg_xch_serialize::ChiaSerialize for NewProofOfSpace {
...
+                &self.message_data,
+                version,
+            ));
+        }
+        if version >= ChiaProtocolVersion::Chia0_0_36 {
+            bytes.extend(dg_xch_serialize::ChiaSerialize::to_bytes(
+                &self.rc_block_unfinished,
+                version,
+            ));
+        }
```
Comment (felixbrucker) [inline core/src/protocols/harvester.rs:205]: Same here

File: core/src/protocols/harvester.rs [hunk 4]
```diff
@@ -211,61 +239,78 @@ impl dg_xch_serialize::ChiaSerialize for RequestSignatures {
...
+                &self.include_source_signature_data,
+                version,
+            ));
+        }
+        if version >= ChiaProtocolVersion::Chia0_0_36 {
+            bytes.extend(dg_xch_serialize::ChiaSerialize::to_bytes(
+                &self.farmer_reward_address_override,
+                version,
+            ));
+        }
```
Comment (felixbrucker) [inline core/src/protocols/harvester.rs:290]: Same here

File: serialize/src/lib.rs [hunk 5]
```diff
@@ -1,25 +1,65 @@
...
 pub fn hash_256(input: impl AsRef<[u8]>) -> Vec<u8> {
     let mut hasher = Sha256::new();
     hasher.update(input);
     hasher.finalize().to_vec()
 }
 
+#[derive(Default, Copy, Clone, Ord, PartialOrd, Eq, PartialEq)]
+pub enum ChiaProtocolVersion {
+    Chia0_0_34 = 34, //Pre 2.0.0
+    #[default]
```
Comment (felixbrucker) [inline serialize/src/lib.rs:18]: Does this mean the default is 35 instead of 36?

File: servers/src/websocket/farmer/handshake.rs [hunk 6]
```diff
@@ -29,23 +30,34 @@ impl MessageHandler for HandshakeHandle {
...
                 .await
                 .send(Message::Binary(
                     ChiaMessage::new(
                         ProtocolMessageTypes::Handshake,
+                        protocol_version,
                         &Handshake {
                             network_id,
-                            protocol_version: PROTOCOL_VERSION.to_string(),
+                            //Server Will use version sent by peer
+                            protocol_version: protocol_version.to_string(),
```
Comment (felixbrucker) [inline servers/src/websocket/farmer/handshake.rs:60]: We should probably only send the peers version if we support it, that being every version below the highest currently implemented one (assuming we support all old versions), but not a version higher

For example:
- peer with 0.0.34 connects, we send back 0.0.34, as we support it
- peer with 0.0.42 connects, we send back 0.0.36, as that's the highest we support (right now)

That way it is clear to the client whether we actually support the version we send, or not

File: servers/src/websocket/farmer/new_proof_or_space.rs [hunk 7]
```diff
@@ -60,13 +60,11 @@ impl<T: PoolClient + Sized + Sync + Send + 'static> MessageHandler for NewProofO
         peer_id: Arc<Bytes32>,
         peers: PeerMap,
     ) -> Result<(), Error> {
-        let exists;
-        {
-            exists = peers.lock().await.get(&peer_id).is_some();
-        }
-        if exists {
+        let peer = peers.read().await.get(&peer_id).cloned();
+        if let Some(peer) = peer {
```
Comment (felixbrucker) [inline servers/src/websocket/farmer/new_proof_or_space.rs:64]: Building christmas trees is hard to read, maybe we can get rid of all this indentation here and below

**Review by felixbrucker — CHANGES_REQUESTED** [review]
Nice work!

File: serialize/src/lib.rs (hunk 5 above)
Comment (DaOneLuna) [reply serialize/src/lib.rs]: It was while the 2.2.0 changes were pending. With the 2.2.1 release I have switched this back

File: servers/src/websocket/farmer/handshake.rs (hunk 6 above)
Comment (DaOneLuna) [reply servers/src/websocket/farmer/handshake.rs:60]: The from_str call above will detect known versions, (0.0.34-36) or will fall back to the default. So this behavior should already be in place.

**Pull request merged**